src/target/riscv/riscv-013.c

```diff
@@ -4284,6 +4288,10 @@ static bool riscv013_is_halted(struct target *target)
 		if (target->state == TARGET_HALTED)
 			dmcontrol |= DM_DMCONTROL_HALTREQ;
 		dmi_write(target, DM_DMCONTROL, dmcontrol);
+
+		RISCV_INFO(r);
+		if (r->on_reset)
+			r->on_reset(target);
```

Comment:
May I ask what the purpose (use case) of this new `on_reset` callback is?

There is no implementation of the callback added. Is a concrete implementation of the callback something you would like to add later?

Also note that OpenOCD already has multiple "on reset" events implemented, for which you can write own TCL callbacks and they will be invoked during target reset. Maybe those events can cover your use cases:

https://openocd.org/doc/html/CPU-Configuration.html#Target-Events

Reply:
I think, reset events are designed to use before/after user triggered resets. (Reset from OpenOCD)
We need that callback to detect reset in case of HW reset button pressed. (Unexpected reset)

In the `on_reset` implementation we are doing some rtos cleaning and displaying the last reset reason.

Reply:
> I think, reset events are designed to use before/after user triggered resets. (Reset from OpenOCD)
> We need that callback to detect reset in case of HW reset button pressed. (Unexpected reset)

Thank you for explaining your use case.

> In the on_reset implementation we are doing some rtos cleaning and displaying the last reset reason.

Two broader points:

(1) The on-reset actions will generally be different for each RISC-V platform - for example reading the reset reason from a platform-specific memory-mapped register. So it would be good to have the "on_reset" event available on the TCL level, so that the users can write their own handlers - just by preparing a custom TCL function. We can't expect an ordinary user to edit the `on_reset` handler directly in OpenOCD source :). 

If you're interested how that is done, you can find an example in `arm_tpiu_swo.c` ([registration of custom event handlers](https://github.com/riscv/riscv-openocd/blob/d97da0eb74330bdeaca0991967451c17f279d062/src/target/arm_tpiu_swo.c#L487), [invocation of a handler](https://github.com/riscv/riscv-openocd/blob/d97da0eb74330bdeaca0991967451c17f279d062/src/target/arm_tpiu_swo.c#L163)).

But that would be a task for a separate merge request (future).

(2) Also note that the on-reset processing may not be reliable unless the `dmcontrol.setresethaltreq` bit is set in the target to halt the target immediately after the reset.

If that bit is not set (or the target does not support it), then the processor will execute an unknown (undetermined) number of instructions before OpenOCD detects the reset, and may be in undetermined state at that point. Or even worse, the  can even reset multiple times before OpenOCD detects that the reset happened.

Reply:
@JanMatCodasip thanks for the valuable points.

For our case it is important to reset FreeRTOS settings when reset happens. We can reset multiple times or whenever OpenOCD detects. We have not seen an issue so far with this implementation. But it is a really good point.

I am open to add features to the more generic layers. Will have a look at your custom event suggestion.